### DISCUSSION TITLE
CSINode does not contain driver driver.longhorn.io

### DISCUSSION THREAD

**Discussion by mqsoh** [longhorn/longhorn#3802] [Q&A] [answered]
I have a cluster whose nodes have taints for particular applications, e.g. people can opt into nodes with a "spot" taint. I installed Longhorn with kubectl and altered the longhorn-manager DaemonSet to tolerate some of those taints. [The manifest has an example of that, commented out.](https://github.com/longhorn/longhorn/blob/v1.2.4/deploy/longhorn.yaml#L1055)

Using the Longhorn frontend I created a ReadWriteMany volume, pv, and pvc. I created three pods with nodeSelectors that select a node with no taints and the other two select nodes with taints. The first pod successfully attaches the volume. The other two have events like:

```
Warning  FailedAttachVolume  2m50s (x48 over 84m)  attachdetach-controller  AttachVolume.Attach failed for volume "sharing-test" : CSINode ip-10-6-62-230.ec2.internal does not contain driver driver.longhorn.io
```

Should there be a `longhorn-csi-plugin` for each manager? Because I have:

```
$ kubectl --namespace=longhorn-system get pods -o=json | jq '.items[].metadata.ownerReferences[].name' -r | grep 'longhorn-manager\|longhorn-csi-plugin' | sort | uniq -c
     30 longhorn-csi-plugin
     64 longhorn-manager
```

The longhorn-csi-plugin DaemonSet isn't in the kubectl installation so I guess there's some Longhorn process that creates it. How do I configure the tolerations for longhorn-csi-plugin?

**Comment by mqsoh** [comment, marked as answer]
I apologize. I was either confused by the nomenclature or possibly burnt out. [The taint-toleration setting is what I need.](https://longhorn.io/docs/1.2.4/references/settings/#kubernetes-taint-toleration) I read that multiple times before posting. 🤷‍♂️

Anyway, this project is awesome and thank you for your work.

**Comment by PhanLe1010** [comment]
@mqsoh Yeah, there should the one Longhorn csi-plugin on every nodes same as the number of Longhorn manager.

I believe that you might have forget to set the toleration setting for Longhorn managed components (longhorn csi-plugin is on of the managed components). 

This document is what you need: https://longhorn.io/docs/1.2.4/advanced-resources/deploy/taint-toleration/#setting-up-taints-and-tolerations-after-longhorn-has-been-installed 

Happy DevOpsing 🙌
